pkg/spanner/migration.go

```diff
@@ -208,6 +225,31 @@ func toStatements(file []byte) ([]string, error) {
 	return statements, nil
 }
 
+func substitutePlaceholders(statements []string, placeholders map[string]string) ([]string, error) {
+	substitutedStatements := []string{}
+	for _, statement := range statements {
+		// Check for any placeholder references that are not configured in the placeholders key value map.
+		placeholdersInStatements := placeholderRegex.FindAllStringSubmatch(statement, -1)
+		for _, match := range placeholdersInStatements {
+			if len(match) == 2 {
+				placeholderInStatement := match[1]
+				_, ok := placeholders[placeholderInStatement]
+				if !ok {
+					return nil, fmt.Errorf("migration statement refers to placeholder variable that is not defined")
+				}
+			}
+		}
```

Comment:
Optional suggestion. The golang regex package is a bit un-ergonomic with these magic indexes based on the match groups. I have an xregexp package in this repo which is a light wrapper around the named groups which lets you write the following.

```suggestion
		m, found := xregexp.FindMatchGroups(placeholderRegex, statement)
		if found {
			_, ok := placeholders[m["PlaceholderName"]]
			if !ok {
				return nil, fmt.Errorf("migration statement refers to placeholder variable that is not defined")
			}
		}
```

With the regex updated to include the named group ``\$\{(?P<PlaceholderName>[A-Za-z_]+)\}``

Reply:
Had to tweak it a bit so it would look for all matches rather than just the left most match. But got there in the end.